packages/core/src/components/event-stream.aggregate-store.ts

```diff
@@ -0,0 +1,174 @@
+  A extends HasTrait<typeof EventSourced> & HasTrait<typeof Identifiable>,
+>(
+  AGGREGATE: A,
+) => {
+  return class $EsAggregateStore extends EventStreamAggregateStore<
+    InstanceType<A>
+  > {
+    loadFirst(event: EventsOf<A>[number]): InstanceType<A> {
+      return AGGREGATE.loadFirst(event);
+    }
+
+    getStreamId(id: InstanceType<A>["id"]): StreamId {
+      return StreamId.from(AGGREGATE.name, id.serialize());
+    }
+  };
+};
+
+export abstract class EventStreamAggregateStore<
+  A extends IEventSourced & IIdentifiable,
+> implements IEsAggregateStore<A>
+{
+  constructor(
+    public readonly streamStore: EventStreamStore<EventOf<A>>,
+    public readonly transaction: TransactionPerformer,
+    public readonly snapshotter: Store<A>,
+  ) {}
+
+  abstract getStreamId(id: A["id"]): StreamId;
+  abstract loadFirst(event: IFact<EventOf<A>>): A;
+
+  async loadFromSnapshot(snapshot: A) {
+    const streamId = this.getStreamId(snapshot.id);
+    const from = snapshot.acknowledgedRevision + 1;
+    const stream = this.streamStore.read(streamId, from);
+
+    for await (const fact of stream) {
+      snapshot.load(fact);
+    }
+
+    return snapshot;
+  }
+
+  async loadFromScratch(id: A["id"]) {
+    const streamId = this.getStreamId(id);
+
+    let instance: A | undefined = undefined;
+    for await (const fact of this.streamStore.read(streamId)) {
+      if (!instance) {
+        instance = this.loadFirst(fact);
+      } else {
+        instance.load(fact);
+      }
+    }
+
+    return instance;
+  }
+
+  async loadForce(id: A["id"]) {
+    const snapshot = await this.snapshotter?.load(id);
+
+    if (snapshot) {
+      return this.loadFromSnapshot(snapshot);
+    }
+
+    return this.loadFromScratch(id);
+  }
+
+  async load(id: A["id"]) {
+    /**
+     * For now, we can just return the snapshot if it exists
+     * because the the snapshot is kept up to date with the event store transactionally.
+     *
+     * Performing the revision check is very expensive when using a query to get the missing events (10x slower)
+     *
+     * In the future, when snapshot will be lagging behind the stream we could:
+     * - Keep a transactionnal revision mutex for each stream, to know if the snapshot is up to date without having to snapshot the whole model
+     * - Use a stale snapshot and let the model catch up with the stream on the next save
+     * - Use a snapshot that is always up to date with the stream, but that is not transactionnal, and catch up with the stream on the next save
+     */
+    const snapshot = await this.snapshotter?.load(id);
+
+    if (snapshot) {
+      return snapshot;
+    }
+
+    return this.loadFromScratch(id);
+  }
+
+  async saveAll(
+    aggregates: A[],
+    parentTrx?: Transaction,
+    attempts = 10,
+  ): Promise<void> {
+    const toSave = await Promise.all(
+      aggregates
+        .filter((a) => a.changes.length)
+        .map(async (aggregate) => ({
```

Comment:
I don't see any await in this async function.

Reply:
fixed!